juju/juju
# minimal jujuc 1867021

### Checklist


## Description of change

This goes back to an old way, and splits out the 'hook-tool' command into an executable called 'jujuc' (which ends up around 8MB when stripped), separated from 'jujud' (116MB stripped). It also updates the symlink build so that if it finds a 'jujuc' it will create the links to it, and updates the tarball to include 'jujuc', etc.

The big win for this, is that 'jujud' takes 60ms to start the process before it realizes that it just wants to send commands to the juju agent, while 'jujuc' takes approx 1ms to get to the same point.

It isn't without cost. There is a price for having the extra 8MB binary:
```
24M Mar 12 10:52 with.tar.xz
21M Mar 12 10:52 without.tar.xz

35M Mar 12 10:52 with.tgz
31M Mar 12 10:52 without.tgz

58M Mar 12 10:50 with-nostrip.tgz
52M Mar 12 10:49 without-nostrip.tgz
```

So it costs around 11-15% to have the extra file in there, regardless of whether we strip or whether we use .xz instead of .gz. However, note that with.tar.xz is still a lot smaller than without.tgz. We could pay back the cost of the extra size if we wanted to go to .xz compression.

I don't actually think this should be in a 2.7 release, but because 2.7 hasn't merged to develop and I prototyped this on 2.7, I wanted to get feedback before rebasing, etc.

This likely makes a huge difference for any charms that call a lot of hook tools (like juju-log). I would really like to see a run of 'juju deploy openstack' with and without this change.

## QA steps

```sh
$ juju bootstrap lxd lxd
$ juju switch controller
$ juju deploy cs:~jameinel/ubuntu-lite --to 0
$ juju ssh 0
$$ ls -l /var/lib/juju/tools/machine-*/
total 53058
-rw-r--r-- 1 ubuntu ubuntu         7 Mar 12 07:35 FORCE-VERSION
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 action-fail -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 action-get -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 action-log -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 action-set -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 add-metric -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 application-version-set -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 close-port -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 config-get -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
lrwxrwxrwx 1 root   root          46 Mar 12 07:38 credential-get -> /var/lib/juju/tools/2.7.5.1-bionic-amd64/jujuc
-rw-r--r-- 1 root   root         209 Mar 12 07:36 downloaded-tools.txt
...
-rwxr-xr-x 1 ubuntu ubuntu   8962048 Mar 12 07:35 jujuc
-rwxr-xr-x 1 ubuntu ubuntu 115477864 Mar 12 07:35 jujud
...
```

You can see that all of those target 'jujuc' instead of 'jujud', and that both 'jujuc' and 'jujud' are installed.

I also added an action to my local copy of ubuntu-lite, which is just:
```
#!/usr/bin/env python3

import subprocess
import time


def count():
    output = subprocess.check_output(['action-get', 'count'])
    return int(output.strip())

def output(val, *extra):
    subprocess.check_call(['action-set', 'output=' + str(val)] + list(extra))


def log(msg):
    #subprocess.check_output(['echo', str(msg), '>/tmp/out.txt'], shell=True)
    subprocess.check_output(['juju-log', str(msg)])


def main():
    c = count()
    start = time.clock_gettime(time.CLOCK_MONOTONIC)
    last = start
    for i in range(c):
        now = time.clock_gettime(time.CLOCK_MONOTONIC)
        log('%8.3fs %8.3fs count: %d' % (now - start, now-last, i))
        # output(last-now)
        # config = subprocess.check_output(['config-get'])
        last = now
    now = time.clock_gettime(time.CLOCK_MONOTONIC)
    total = now - start
    # output(str(total / c), 'config='+str(config))
    output(str(total / c))


if __name__ == '__main__':
    main()
```

Without this patch, on lxd on my machine, I see it taking ~60ms per call to 'juju-log'. With the patch, it drops down to around 6ms.

Note that `juju bootstrap` with prebuilt binaries and `juju upgrade-controller --build-agent` use slightly different codepaths, so it can be worthwhile to test both. I also haven't vetted the full release process, to make sure that snaps will have all the binaries we need, nor that the process that generates streams.canonical.com would do the right thing. 

with the patch:
```
$ juju run-action --wait ubuntu-lite/0 logging-test
unit-ubuntu-lite-0:
  UnitId: ubuntu-lite/0
  id: "1"
  results:
    output: "0.006217624710407108"
```
without the patch:
```
$ juju run-action --wait ubuntu-lite/0 logging-test
unit-ubuntu-lite-0:
  UnitId: ubuntu-lite/0
  id: "1"
  results:
    output: "0.07029640914988704"
```



## Documentation changes

Nothing changes for the user workflow, though it should make install and various hooks quite a bit faster. Especially reactive charms that spend a lot of time in 'juju-log'.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1867021
